TFNetworkLayer.py

```diff
@@ -1107,53 +1111,218 @@ def batch_norm(self, data,
+      bn = (input_data.placeholder - mean) * tf.rsqrt(variance + epsilon)
+      if use_std:
+        bn *= gamma
+      if use_shift:
         bn += beta
+      if data.is_batch_feature_major and not masked_time:
+        # Transpose back to NCHW
+        perm = data.get_axes()
+        perm = perm[:1] + perm[-1:] + perm[1:-1]
+        bn = tf.transpose(bn, perm=perm)
       return bn
 
+  def _fused_batch_norm(self, data,
+                        axes,
+                        use_shift, use_std,
+                        momentum, epsilon,
+                        sample_mean, sample_variance,
+                        gamma, beta,
+                        masked_time):
+    """
+    :param Data data:
+    :param int|str|list[int|str] axes: axis/axes to normalize
+    :param bool use_shift:
+    :param bool use_std:
+    :param float momentum: for the running average of sample_mean and sample_std
+    :param float epsilon:
+    :param tf.Variable sample_mean:
+    :param tf.Variable sample_variance:
+    :param tf.Variable gamma:
+    :param tf.Variable beta:
+    :param bool masked_time: flatten and mask input tensor
+    :rtype: tf.Tensor
+
+    See tf.nn.fused_batch_norm()
+    """
+    assert sample_mean is not None
+    assert sample_variance is not None
+    assert not use_std or gamma is not None
+    assert not use_shift or beta is not None
+
+    # with tf.variable_scope("fused_batch_norm"):
+    axes = data.get_axes_from_description(axes)
+    assert data.feature_dim_axis in axes
+    assert data.batch_dim_axis not in axes  # Not supported
+    assert data.time_dim_axis not in axes  # Not supported
+    if masked_time:
+      assert len(axes) == 1 and axes[0] == data.feature_dim_axis
+      x = data.get_placeholder_flattened(keep_dims=True)
+    else:
+      # Put feature_dim_axis at the end s.t. the merged axis is a feature.
+      axes.remove(data.feature_dim_axis)
+      axes.append(data.feature_dim_axis)
+      x = data.get_placeholder_merge_axes(axes=axes, keep_dims=True)
+    data_format = "NCHW" if data.is_batch_feature_major else "NHWC"
+
+    def _fused_batch_norm_training():
+      return tf.nn.fused_batch_norm(
+        x,
+        gamma,
+        beta,
+        epsilon=epsilon,
+        data_format=data_format)
+
+    def _fused_batch_norm_inference():
+      return tf.nn.fused_batch_norm(
+        x,
+        gamma,
+        beta,
+        mean=sample_mean,
+        variance=sample_variance,
+        epsilon=epsilon,
+        is_training=False,
+        data_format=data_format)
+
+    output, mean, variance = tf.cond(
+      tf.cast(self.network.train_flag, tf.bool),
+      true_fn=_fused_batch_norm_training,
+      false_fn=_fused_batch_norm_inference)
+    # Remove Bessel's correction to be consistent with non-fused batch norm.
+    sample_size = tf.cast(
+      tf.size(x) / tf.size(variance), variance.dtype)
+    factor = (sample_size - tf.cast(1.0, variance.dtype)) / sample_size
+    variance *= factor
```

Comment:
So the fused variant is actually different than what we have?
That is good to know.
But instead of just undoing what they do, maybe what they do actually makes sense? You should not just undo it. If that is actually the better way to do it (using Bessel's correction), then we also should have that (have an option for it).